parsec/api/data/base.py

```diff
@@ -0,0 +1,203 @@
+# Parsec Cloud (https://parsec.cloud) Copyright (c) AGPLv3 2019 Scille SAS
+
+import attr
+from typing import Optional, Union
+from pendulum import Pendulum
+from nacl.exceptions import CryptoError
+from nacl.secret import SecretBox
+from nacl.public import SealedBox
+from nacl.bindings import crypto_sign_BYTES
+
+from parsec.types import DeviceID
+from parsec.serde import BaseSchema, fields, SerdeValidationError, SerdePackingError, Serializer
+from parsec.crypto_types import PrivateKey, PublicKey, SigningKey, VerifyKey
+
+
+class DataError(Exception):
+    pass
+
+
+class DataValidationError(SerdeValidationError, DataError):
+    pass
+
+
+class DataSerializationError(SerdePackingError, DataError):
+    pass
+
+
+class BaseSignedDataSchema(BaseSchema):
+    author = fields.DeviceID(missing=None)
+    timestamp = fields.DateTime(required=True)
+
+
+class SignedDataMeta(type):
+    CLS_ATTR_COOKING = attr.s(slots=True, frozen=True, auto_attribs=True, kw_only=True)
+
+    def __new__(cls, name, bases, nmspc):
+        if "SCHEMA_CLS" not in nmspc:
+            raise RuntimeError("Missing attribute `SCHEMA_CLS` in class definition")
+        if not issubclass(nmspc["SCHEMA_CLS"], BaseSignedDataSchema):
+            raise RuntimeError(f"Attribute `SCHEMA_CLS` must inherit {BaseSignedDataSchema!r}")
+
+        if "__attrs_attrs__" not in nmspc:
+            if "SERIALIZER" in nmspc and bases:
+                raise RuntimeError("Attribute `SERIALIZER` is reserved")
+            nmspc["SERIALIZER"] = Serializer(
+                nmspc["SCHEMA_CLS"], DataValidationError, DataSerializationError
+            )
+            raw_cls = type.__new__(cls, name, bases, nmspc)
+            return cls.CLS_ATTR_COOKING(raw_cls)
+        else:
+            return type.__new__(cls, name, bases, nmspc)
+
+
+class BaseSignedData(metaclass=SignedDataMeta):
+    SCHEMA_CLS = BaseSignedDataSchema  # Must be overloaded by child class
+    # SERIALIZER attribute sets by the metaclass
```

Comment:
Documenting this class seems quite important since it embeds the meta class and all of its fancy behavior.

Reply:
done